snap_http/http.py

```diff
@@ -15,39 +16,49 @@ class SnapdHttpException(Exception):
     """An exception raised during HTTP communication with snapd."""
 
 
-def get(path: str) -> SnapdResponse:
+def get(
+    path: str,
+    query_params: Optional[Dict[str, Any]] = None,
+) -> SnapdResponse:
     """Peform a GET request of `path`."""
-    response = _make_request(path, "GET")
+    response = _make_request(path, "GET", query_params=query_params)
 
     return SnapdResponse.from_http_response(response)
 
 
 def post(path: str, body: SnapdRequestBody) -> SnapdResponse:
     """Perform a POST request of `path`, JSON-ifying `body`."""
-    response = _make_request(path, "POST", body)
+    response = _make_request(path, "POST", body=body)
 
     return SnapdResponse.from_http_response(response)
 
 
 def put(path: str, body: SnapdRequestBody) -> SnapdResponse:
     """Perform a PUT request of `path`, JSON-ifying `body`."""
-    response = _make_request(path, "PUT", body)
+    response = _make_request(path, "PUT", body=body)
 
     return SnapdResponse.from_http_response(response)
 
 
 def _make_request(
-    path: str, method: str, body: Optional[SnapdRequestBody] = None
+    path: str,
+    method: str,
+    *,
+    body: Optional[SnapdRequestBody] = None,
+    query_params: Optional[Dict[str, Any]] = None,
 ) -> Any:
     """Performs a request to `path` using `method`, including `body`, if provided.
 
     urllib doesn't support HTTP requests to UNIX sockets, so we create out own socket, start the
     connection, then hand it off to `HTTPResponse` to read from and parse.
     """
     sock = socket.socket(family=socket.AF_UNIX)
-
     sock.connect(SNAPD_SOCKET)
+
     url = BASE_URL + path
+    if query_params:
+        url += "?" + urlencode(query_params)
```

Comment:
We may need to be careful here – there's two ways that spaces and `/` can be encoded, and the default for python is `+` and `%2F`, respectively, but one alternative is `%20` and not changing `/`. Let's make sure we have an integration test where we see the expected result out of snapd when we set configs that contain these characters.

Reply:
hmm actually after some research it appears that you cannot use spaces or `/` in a snap config value key, so we only need to worry about this if that changes. Maybe an integration test for a failed `set_conf` using invalid chars?

Reply:
Added the integration test for `set_conf` with invalid chars